config/package_resouce.go

```diff
@@ -14,26 +14,229 @@
+		case agentendpointpb.ApplyConfigTask_Config_Resource_PackageResource_INSTALLED:
+			p.policy.GooGet.Install = pr
+		case agentendpointpb.ApplyConfigTask_Config_Resource_PackageResource_REMOVED:
+			p.policy.GooGet.Remove = pr
+		}
+	case *agentendpointpb.ApplyConfigTask_Config_Resource_PackageResource_Yum:
+		pr := p.GetYum()
+		if !packages.YumExists {
+			return nil, fmt.Errorf("cannot manage Yum package %q because yum does not exist on the system", pr.GetName())
+		}
+
+		switch p.GetDesiredState() {
+		case agentendpointpb.ApplyConfigTask_Config_Resource_PackageResource_INSTALLED:
+			p.policy.Yum.Install = pr
+		case agentendpointpb.ApplyConfigTask_Config_Resource_PackageResource_REMOVED:
+			p.policy.Yum.Remove = pr
+		}
+	case *agentendpointpb.ApplyConfigTask_Config_Resource_PackageResource_Zypper_:
+		pr := p.GetZypper()
+		if !packages.ZypperExists {
+			return nil, fmt.Errorf("cannot manage Zypper package %q because zypper does not exist on the system", pr.GetName())
+		}
+
+		switch p.GetDesiredState() {
+		case agentendpointpb.ApplyConfigTask_Config_Resource_PackageResource_INSTALLED:
+			p.policy.Zypper.Install = pr
+		case agentendpointpb.ApplyConfigTask_Config_Resource_PackageResource_REMOVED:
+			p.policy.Zypper.Remove = pr
+		}
+	case *agentendpointpb.ApplyConfigTask_Config_Resource_PackageResource_Rpm:
+		pr := p.GetRpm()
+		if !packages.RPMExists {
+			return nil, fmt.Errorf("cannot manage RPM package because rpm does not exist on the system")
+		}
+
+		switch p.GetDesiredState() {
+		case agentendpointpb.ApplyConfigTask_Config_Resource_PackageResource_INSTALLED:
+			p.policy.RPM.Install = pr
+		case agentendpointpb.ApplyConfigTask_Config_Resource_PackageResource_REMOVED:
+			p.policy.RPM.Remove = pr
+		}
+	default:
+		return nil, errors.New("SystemPackage field not set or references unknown package manager")
+	}
+
+	return &ManagedResources{Packages: []ManagedPackage{p.policy}}, nil
+}
+
+// TODO: implement a caching system for installed packages.
+var aptInstalled []packages.PkgInfo
+var debInstalled []packages.PkgInfo
+var gooInstalled []packages.PkgInfo
+var yumInstalled []packages.PkgInfo
+var zypperInstalled []packages.PkgInfo
+var rpmInstalled []packages.PkgInfo
+
+func needsInstall(installedPkgs []packages.PkgInfo, name string) bool {
+	for _, pkg := range installedPkgs {
+		if pkg.Name == name {
+			return false
+		}
+	}
+	return true
 }
 
 func (p *packageResouce) checkState() (bool, error) {
-	// TODO: implement
-	return true, nil
+	switch {
+	case p.policy.Apt.Install != nil:
+		if needsInstall(aptInstalled, p.policy.Apt.Install.GetName()) {
+			return true, nil
+		}
+	case p.policy.Apt.Remove != nil:
+		if !needsInstall(aptInstalled, p.policy.Apt.Remove.GetName()) {
+			return true, nil
+		}
+
+	// TODO: implement check for deb
+	case p.policy.Deb.Install != nil:
+	case p.policy.Deb.Remove != nil:
+
+	case p.policy.GooGet.Install != nil:
+		if needsInstall(aptInstalled, p.policy.GooGet.Install.GetName()) {
```

Comment:
gooInstalled?

Reply:
needsInstall is the function name, renamed to be clearer

Reply:
Sorry, I was referring to the argument passed to _needsInstall_ function. In this case, it should be _gooInstalled_, correct?

Reply:
oh lol, yeah nice catch, I will have full unit tests for this once it's fully implemented.